python/ray/data/dataset.py

```diff
@@ -3709,13 +3709,13 @@ def write_datasink(
             datasink.on_write_start()
 
             self._write_ds = Dataset(plan, logical_plan).materialize()
-            blocks = ray.get(self._write_ds._plan.execute().block_refs)
+            raw_write_results = ray.get(self._write_ds._plan.execute().block_refs)
```

Comment:
(not a new issue from this PR, can be done later if needed)
I think we should get and handle the blocks with an iterator, instead of getting everything in a blocking way.  
So some blocks can be freed earlier.